# Introduction to Data Analyst
Dalam notebook ini anda akan diajarkan mengenai data pre-processing, salah satu langkah penting dalam mengolah data.

## Connect Gdrive kedalam Colab

In [ ]:
# Mengakses google drive ke dalam google colaboratory
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


## Import Packages
Packages/library adalah salah satu tools yang kedepannya akan memudahkan anda dalam mengeksekusi data anda. Dengan library anda tidak membutuhkan code manual dan dapat menggunakan function atau method yang terdapat dalam libabry tersebut.

In [ ]:
# Import library
import pandas as pd
from scipy.stats import mode
import os

## Import Data
Berikut akan dicontohkan cara membaca file dengan format yang berbeda dan menampilkannya dalam bentuk dataframe.

### Read CSV File
Langkah pertama dalam membaca file adalah memastikan dimana file anda tersimpan. Buatlah path yang merepresentasikan letak data anda tersimpan.

In [ ]:
# Inisiasi directory tempat data akan dibaca dan disimpan
path = "gdrive/MyDrive/Mini Course/DATA/" #silahkan masukkan path sesuai dengan letak data di drive masing-masing

In [ ]:
os.listdir(path)

['Provinsi Jawa Timur Dalam Angka 2020-pages-85.pdf',
 'OnlineRetail.xlsx',
 'OnlineRetail.csv']

Langkah ke-dua adalah membaca data. Untuk membaca data dan menampilkannya dalam bentuk dataframe anda akan menggunakan package yaitu pandas. Dataframe banyak digunakan oleh data analyst dalam mengolah data. Salah satu faktor utama penggunaan dataframe adalah data terdisplay dengan schema yang jelas sehingga memudahkan pembacaan data. Selain itu perubahan-perubahan yang akan anda lakukan dapat anda amati secara langsung dan memastikan apakah sudah sesuai dengan apa yang anda harapkan.

In [ ]:
# Membaca file csv
df = pd.read_csv(path + 'OnlineRetail.csv', encoding = "ISO-8859-1")

In [ ]:
# Menampilkan 5 urutan teratas dari data
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

Anda telah berhasil membaca data dalam format csv file, selanjutkan akan dicontohkan cara membaca data dengan format xlsx.

### Read Excel File

Sama seperti langkah sebelumnya, dalam membaca data excel akan digunakan pandas. Yang membedakan antara membaca data csv dan excel adalah perintah pembacaan datanya. Pada bagian "read_(isi_formatnya)" setiap file akan membutuhkan cara pembacaan yang berbeda. Untuk file excel anda dapat memasukkan code dibawah ini.

In [ ]:
# Membaca file excel
df_xlsx = pd.read_excel(path + "OnlineRetail.xlsx")

In [ ]:
# Menampilkan 5 urutan teratas dari data
df_xlsx.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

Sekarang anda telah berhasil membaca 2 data yang berbeda, untuk kedepannya akan digunakan csv data untuk step-step berikutnya.

> *Sebagai catatan jika anda ingin mencoba pada data excel anda, anda tetap dapat melakukannya seperti langkah-langkah dibawah ini. Perbedaan data format data hanya berbeda di proses awal pembacaan datanya saja.*

## Exploratory Data Analysis (EDA)
Secara garis besar EDA adalah suatu proses dalam menganalis data yang anda miliki, contohnya melihat pola data, informasi data, menemukan anomali dari data, dan beberapa analisis statistik lainnya.

### Data terdiri dari berapa baris dan berapa kolom ???

In [ ]:
# Mencari ukuran data
df.shape

(541909, 8)

Cara membaca output tersebut adalah (baris, kolom). Dari data diatas maka diketahui bahwa data tersebut terdiri dari **541909 baris** dan **8 kolom**.

### Bagaimana kondisi data nya? bermasalah ngga?


In [ ]:
# Melihat Informasi Data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Data informasi berisi mengenai jumlah baris dalam data, nama-nama kolom dan jumlah element dari kolom tersebut, serta tipe data dari masing-masing kolom. Mengetahui informasi dari data akan memudahkan anda untuk mencari tahu apa saja yang perlu anda lakukan seperti merubah format data, melakukan manipulasi dalam data.

## Pre-Processing Data
Dalam tahap ini anda akan mencoba untuk cleaning dan manipulate data anda.

### Apakah ada Missing value ?

In [ ]:
# Cek missing value
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Berdasarkan output diatas diketahui bahwa terdapat missing value pada kolom 'Description' dan 'CustomerID', oleh karena itu perlu dilakukan cleaning pada kolom tersebut. Salah satu cara dalam handling missing value adalah drop data yang miss.

In [ ]:
# Handling missing value dengan drop data
df = df.dropna(axis=0)

# Cek kembali missing value
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   InvoiceDate  406829 non-null  object 
 5   UnitPrice    406829 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      406829 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 27.9+ MB


### Apakah ada Data Duplicate ?

Mengecek data yang duplikat adalah salah satu proses penting dalam cleaning data. Duplikat data akan menjadikan analisa menjadi kurang valid dikarenakan terlalu banyak data yang seragam.

In [ ]:
# Mencari duplicate data
duplicate = df[df.duplicated()]

# Display data yang duplicate
duplicate

InvoiceNo StockCode  ... CustomerID         Country
517       536409     21866  ...    17908.0  United Kingdom
527       536409     22866  ...    17908.0  United Kingdom
537       536409     22900  ...    17908.0  United Kingdom
539       536409     22111  ...    17908.0  United Kingdom
555       536412     22327  ...    17920.0  United Kingdom
...          ...       ...  ...        ...             ...
541675    581538     22068  ...    14446.0  United Kingdom
541689    581538     23318  ...    14446.0  United Kingdom
541692    581538     22992  ...    14446.0  United Kingdom
541699    581538     22694  ...    14446.0  United Kingdom
541701    581538     23343  ...    14446.0  United Kingdom

[5225 rows x 8 columns]

In [ ]:
# Handling duplicate data
df = df.drop_duplicates()

In [ ]:
df

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...      17850  United Kingdom
1         536365     71053  ...      17850  United Kingdom
2         536365    84406B  ...      17850  United Kingdom
3         536365    84029G  ...      17850  United Kingdom
4         536365    84029E  ...      17850  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...      12680          France
541905    581587     22899  ...      12680          France
541906    581587     23254  ...      12680          France
541907    581587     23255  ...      12680          France
541908    581587     22138  ...      12680          France

[401604 rows x 8 columns]

### Ubah Tipe data
Dalam beberapa kasus terkadang dijumpai beberapa data yang memiliki tpe data yang kurang sesuai dengan yang dibutuhkan. Oleh karena itu pengubahan tipe data sangat diperlukan.

In [ ]:
# Mengubah tipe data 'InvoiceDate' kedalam bentuk Datetime
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [ ]:
# Mengubah tipe data 'CustomerID' kedalam bentuk Non-Metric atau string
df['CustomerID'] = df['CustomerID'].astype(int).astype(str)

In [ ]:
# Mengecek kembali informasi dari data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  object        
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 27.6+ MB


In [ ]:
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...      17850  United Kingdom
1    536365     71053  ...      17850  United Kingdom
2    536365    84406B  ...      17850  United Kingdom
3    536365    84029G  ...      17850  United Kingdom
4    536365    84029E  ...      17850  United Kingdom

[5 rows x 8 columns]

Berdasarkan informasi data, dapat dilihat bahwa tipe data sudah berubah sesuai yang diharapkan dan semua kolom memiliki ukuran yang sama.

### Melakukan Analisa

#### Apa saja variabel yang termasuk dalam non-metric ?

In [ ]:
# Menampilkan kolom/variabel dengan tipe data non-metric atau string
non_metric = list(df.select_dtypes(include=['object']).columns)
df[non_metric].head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...      17850  United Kingdom
1    536365     71053  ...      17850  United Kingdom
2    536365    84406B  ...      17850  United Kingdom
3    536365    84029G  ...      17850  United Kingdom
4    536365    84029E  ...      17850  United Kingdom

[5 rows x 5 columns]

#### Berapakah jumlah nilai yang unik pada setiap variabel ?

In [ ]:
df[non_metric].agg(['nunique','count'])

InvoiceNo  StockCode  Description  CustomerID  Country
nunique      22190       3684         3896        4372       37
count       406829     406829       406829      406829   406829

Index nunique menunjukkan nilai unique dari tiap-tiap kolom, sedangkan count menunjukkan total keseluruhan data dalam kolom.

#### Informasi apa yang paling banyak muncul dari setiap variabel secara keseluruhan ?

In [ ]:
# Mencari modus dari data non-metric
df[non_metric].mode()

InvoiceNo StockCode  ... CustomerID         Country
0    576339    85123A  ...      17841  United Kingdom

[1 rows x 5 columns]

#### Informasi apa yang paling banyak muncul dan berapakah jumlah nilai yang unik dari setiap variabel pada setiap negara ?

In [ ]:
# Membuat analisa berdasarkan country dan mencari tahu informasi seperti unique value, total data, dan modus pada setiap variabel
df.groupby('Country').agg(['nunique','count',pd.Series.mode]).stack()

InvoiceNo       StockCode  ... UnitPrice CustomerID
Country                                           ...                     
Australia      nunique        69             600  ...        77          9
               count        1258            1258  ...      1258       1258
               mode       556917           22720  ...      1.65      12415
Austria        nunique        19             307  ...        44         11
               count         401             401  ...       401        401
...                          ...             ...  ...       ...        ...
United Kingdom count      356728          356728  ...    356728     356728
               mode       576339          85123A  ...      1.25      17841
Unspecified    nunique         8             213  ...        44          4
               count         241             241  ...       241        241
               mode       561658  [23108, 84879]  ...      1.25      12743

[111 rows x 6 columns]

#### Berapa rata-rata, nilai minimum, nilai maksimum kuantitas barang yang terujual dan harganya ? 

In [ ]:
# Melihat summary dari data yang telah di cleaning
df.describe()

Quantity      UnitPrice
count  406829.000000  406829.000000
mean       12.061303       3.460471
std       248.693370      69.315162
min    -80995.000000       0.000000
25%         2.000000       1.250000
50%         5.000000       1.950000
75%        12.000000       3.750000
max     80995.000000   38970.000000

### Menambahkan Variabel Baru
Dalam menganalisa data terkadang kita dapat menemukan informasi tambahan dari data yang kita miliki. Misalnya terkadang kita tidak mengetahui usia dari user, yang kita hanya miliki tanggal lahir dari user. Oleh karena itu diperlukan variabel/kolom tambahan untuk mengidentifikasi usia dari user tersebut dengan memanfaatkan data yang kita miliki saat ini. Langkah seperti ini juga akan kita lakukan pada step ini, yaitu menambah variabel baru berupa keterangan waktu untuk memudahkan analisa kedepannya.

In [ ]:
# Menambahkan kolom baru 
df['year'], df['month'], df['day'] = df['InvoiceDate'].dt.year, df['InvoiceDate'].dt.month, df['InvoiceDate'].dt.day

In [ ]:
# Menampilkan 5 urutan terakhir dari data
df.tail()

InvoiceNo StockCode                      Description  ...  year month  day
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS  ...  2011     9   12
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL   ...  2011     9   12
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL   ...  2011     9   12
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE  ...  2011     9   12
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT   ...  2011     9   12

[5 rows x 11 columns]

## Exporting Data
Berikut merupakan cara untuk menyimpan data hasil pre-processing 

In [ ]:
# Export data kedalam format csv
df.to_csv(path + 'Clean_OnlineRetail.csv', index=None) # keterangan index dalam data ini tidak akan disimpan 

In [ ]:
# Export data kedalam format excel
df.to_excel(path + 'Clean_OnlineRetail.xlsx', index=None)